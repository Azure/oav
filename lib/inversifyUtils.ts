import { Container, interfaces } from "inversify";
import { setDefaultOpts } from "./swagger/loader";

export const TYPES = {
  opts: Symbol("InversifyTYPES.opts"),
};

export const inversifyGetInstance = <T, Opt = {}>(
  claz: interfaces.Newable<T>,
  opts: Opt &
    interfaces.ContainerOptions & {
      container?: Container;
    }
) => {
  if (opts.container === undefined) {
    setDefaultOpts(opts, {
      defaultScope: "Singleton",
      autoBindInjectable: true,
    } as any);
    opts.container = new Container(opts);
  }
  opts.container.bind(TYPES.opts).toConstantValue(opts);
  return opts.container.get(claz);
};
