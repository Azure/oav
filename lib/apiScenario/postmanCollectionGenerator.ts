import * as path from "path";
import * as fs from "fs";
import newman, { NewmanRunOptions, NewmanRunSummary } from "newman";
import { inject, injectable } from "inversify";
import { Collection, VariableScope } from "postman-collection";
import { inversifyGetInstance, TYPES } from "../inversifyUtils";
import { ReportGenerator as HtmlReportGenerator } from "../report/generateReport";
import { FileLoader } from "../swagger/fileLoader";
import { getApiVersionFromFilePath, getRandomString } from "../util/utils";
import {
  OperationCoverageInfo,
  RuntimeException,
  TrafficValidationIssue,
  TrafficValidationOptions,
  unCoveredOperationsFormat,
} from "../swaggerValidator/trafficValidator";
import { LiveValidationIssue } from "../liveValidation/liveValidator";
import { logger } from "./logger";
import { ApiScenarioLoader, ApiScenarioLoaderOption } from "./apiScenarioLoader";
import { ApiScenarioRunner } from "./apiScenarioRunner";
import { generateMarkdownReportHeader } from "./markdownReport";
import { PostmanCollectionRunnerClient } from "./postmanCollectionRunnerClient";
import {
  ApiScenarioTestResult,
  NewmanReportValidator,
  NewmanReportValidatorOption,
} from "./newmanReportValidator";
import { SwaggerAnalyzer, SwaggerAnalyzerOption } from "./swaggerAnalyzer";
import { EnvironmentVariables } from "./variableEnv";
import { parseNewmanSummary } from "./newmanReportParser";
import {
  defaultCollectionFileName,
  defaultEnvFileName,
  defaultNewmanDir,
  defaultNewmanReport,
  defaultQualityReportFilePath,
} from "./defaultNaming";
import { DataMasker } from "./dataMasker";
import { Scenario, ScenarioDefinition } from "./apiScenarioTypes";
import { CLEANUP_FOLDER, PREPARE_FOLDER } from "./postmanHelper";

export interface PostmanCollectionGeneratorOption
  extends ApiScenarioLoaderOption,
    SwaggerAnalyzerOption {
  name: string;
  fileRoot: string;
  scenarioDef: string;
  env: EnvironmentVariables;
  outputFolder: string;
  markdown?: boolean;
  junit?: boolean;
  html?: boolean;
  runCollection: boolean;
  generateCollection: boolean;
  testProxy?: string;
  skipValidation?: boolean;
  savePayload?: boolean;
  generateExample?: boolean;
  skipCleanUp?: boolean;
  runId?: string;
  verbose?: boolean;
  devMode?: boolean;
}

export const generateRunId = (): string => {
  const today = new Date();
  const yyyy = today.getFullYear().toString();
  const MM = pad(today.getMonth() + 1, 2);
  const dd = pad(today.getDate(), 2);
  const hh = pad(today.getHours(), 2);
  const mm = pad(today.getMinutes(), 2);
  const id = getRandomString();
  return yyyy + MM + dd + hh + mm + "-" + id;
};

function pad(number: number, length: number) {
  let str = "" + number;
  while (str.length < length) {
    str = "0" + str;
  }
  return str;
}

@injectable()
export class PostmanCollectionGenerator {
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(
    @inject(TYPES.opts) private opt: PostmanCollectionGeneratorOption,
    private apiScenarioLoader: ApiScenarioLoader,
    private fileLoader: FileLoader,
    private dataMasker: DataMasker,
    private swaggerAnalyzer: SwaggerAnalyzer
  ) {}

  public async run(): Promise<Collection> {
    const scenarioDef = await this.apiScenarioLoader.load(this.opt.scenarioDef);

    await this.swaggerAnalyzer.initialize();
    for (const it of scenarioDef.requiredVariables) {
      if (this.opt.env[it] === undefined) {
        throw new Error(
          `Missing required variable '${it}', please set variable values in env.json.`
        );
      }
    }
    this.opt.runId = this.opt.runId || generateRunId();

    if (this.opt.markdown) {
      const reportExportPath = path.resolve(
        this.opt.outputFolder,
        `${defaultNewmanDir(this.opt.name, this.opt.runId!)}`
      );
      await this.fileLoader.writeFile(
        path.join(reportExportPath, "report.md"),
        generateMarkdownReportHeader()
      );
    }

    const client = new PostmanCollectionRunnerClient({
      collectionName: scenarioDef.name,
      runId: this.opt.runId,
      testProxy: this.opt.testProxy,
      verbose: this.opt.verbose,
      skipAuth: this.opt.devMode,
      skipArmCall: this.opt.devMode,
      skipLroPoll: this.opt.devMode,
    });
    const runner = new ApiScenarioRunner({
      jsonLoader: this.apiScenarioLoader.jsonLoader,
      env: this.opt.env,
      client: client,
    });

    await runner.execute(scenarioDef);

    let [collection, environment] = client.outputCollection();

    if (this.opt.generateCollection) {
      await this.writeCollectionToJson(scenarioDef.name, collection, environment);
    }

    if (this.opt.runCollection) {
      if (collection.items.find((item) => item.name === PREPARE_FOLDER, collection)) {
        const summary = await this.runCollection({
          collection,
          environment,
          folder: PREPARE_FOLDER,
          reporters: "cli",
        });
        environment = summary.environment;
      }
      for (const scenario of scenarioDef.scenarios) {
        const reportExportPath = path.resolve(
          this.opt.outputFolder,
          `${defaultNewmanReport(this.opt.name, this.opt.runId!, scenario.scenario)}`
        );
        const summary = await this.runCollection({
          collection,
          environment,
          folder: scenario.scenario,
          reporters: "cli",
        });
        await this.postRun(scenario, reportExportPath, summary.environment, summary);
        environment = summary.environment;
      }
      if (collection.items.find((item) => item.name === CLEANUP_FOLDER, collection)) {
        if (!this.opt.skipCleanUp) {
          await this.runCollection({
            collection,
            environment,
            folder: CLEANUP_FOLDER,
            reporters: "cli",
          });
        }
      }
    }

    const operationIdCoverageResult = this.swaggerAnalyzer.calculateOperationCoverage(scenarioDef);
    logger.info(
      `Operation coverage ${(operationIdCoverageResult.coverage * 100).toFixed(2) + "%"} (${
        operationIdCoverageResult.coveredOperationNumber
      }/${operationIdCoverageResult.totalOperationNumber})`
    );
    if (operationIdCoverageResult.uncoveredOperationIds.length > 0) {
      logger.verbose("Uncovered operationIds: ");
      logger.verbose(operationIdCoverageResult.uncoveredOperationIds);
    }

    if (this.opt.html && this.opt.runCollection) {
      await this.generateHtmlReport(scenarioDef);
    }

    return collection;
  }

  private async generateHtmlReport(scenarioDef: ScenarioDefinition) {
    const trafficValidationResult = new Array<TrafficValidationIssue>();
    const reportExportPath = path.resolve(
      this.opt.outputFolder,
      `${defaultNewmanDir(this.opt.name, this.opt.runId!)}`
    );

    let providerNamespace;

    for (const dir of fs.readdirSync(reportExportPath, { withFileTypes: true })) {
      if (dir.isDirectory()) {
        const report = JSON.parse(
          await this.fileLoader.load(path.join(reportExportPath, dir.name, "report.json"))
        ) as ApiScenarioTestResult;

        providerNamespace = report.providerNamespace;
        for (const step of report.stepResult) {
          const trafficValidationIssue: TrafficValidationIssue = {
            errors: [
              ...(step.liveValidationResult?.requestValidationResult.errors ?? []),
              ...(step.liveValidationResult?.responseValidationResult.errors ?? []),
            ],
            specFilePath: step.specFilePath,
            operationInfo: step.liveValidationResult?.requestValidationResult.operationInfo ?? {
              operationId: "unknown",
              apiVersion: "unknown",
            },
          };

          // mock
          trafficValidationIssue.operationInfo!.position = {
            line: 0,
            column: 0,
          };

          if (this.opt.savePayload) {
            const payloadFilePath = path.join(
              ".",
              dir.name,
              `payloads/${step.stepName}_${step.correlationId}.json`
            );
            trafficValidationIssue.payloadFilePath = payloadFilePath;
          }

          for (const runtimeError of step.runtimeError ?? []) {
            trafficValidationIssue.errors?.push(this.convertRuntimeException(runtimeError));
          }

          if (step.liveValidationResult?.requestValidationResult.runtimeException) {
            trafficValidationIssue.errors?.push(
              this.convertRuntimeException(
                step.liveValidationResult!.requestValidationResult.runtimeException
              )
            );
          }

          if (step.liveValidationResult?.responseValidationResult.runtimeException) {
            trafficValidationIssue.errors?.push(
              this.convertRuntimeException(
                step.liveValidationResult!.responseValidationResult.runtimeException
              )
            );
          }

          trafficValidationResult.push(trafficValidationIssue);
        }
      }
    }

    const operationIdCoverageResult =
      this.swaggerAnalyzer.calculateOperationCoverageBySpec(scenarioDef);

    const operationCoverageResult: OperationCoverageInfo[] = [];
    operationIdCoverageResult.forEach((result, key) => {
      let specPath = this.fileLoader.resolvePath(key);
      if (process.env.REPORT_SPEC_PATH_PREFIX) {
        specPath = path.join(
          process.env.REPORT_SPEC_PATH_PREFIX,
          specPath.substring(specPath.indexOf("specification"))
        );
      }
      operationCoverageResult.push({
        totalOperations: result.totalOperationNumber,
        spec: specPath,
        coverageRate: result.coverage,
        apiVersion: getApiVersionFromFilePath(specPath),
        unCoveredOperations: result.uncoveredOperationIds.length,
        coveredOperaions: result.totalOperationNumber - result.uncoveredOperationIds.length,
        validationFailOperations: trafficValidationResult.filter(
          (it) => key.indexOf(it.specFilePath!) !== -1 && it.errors!.length > 0
        ).length,
        unCoveredOperationsList: result.uncoveredOperationIds.map((id) => {
          return { operationId: id };
        }),
        unCoveredOperationsListGen: Object.values(
          result.uncoveredOperationIds
            .map((id) => {
              return { operationId: id, key: id.split("_")[0] };
            })
            .reduce((res: { [key: string]: unCoveredOperationsFormat }, item) => {
              /* eslint-disable no-unused-expressions */
              res[item.key]
                ? res[item.key].operationIdList.push(item)
                : (res[item.key] = {
                    operationIdList: [item],
                  });
              /* eslint-enable no-unused-expressions */
              return res;
            }, {})
        ),
      });
    });

    const options: TrafficValidationOptions = {
      reportPath: path.resolve(reportExportPath, "report.html"),
      overrideLinkInReport: false,
      sdkPackage: providerNamespace,
      markdownPath: this.opt.markdown ? path.resolve(reportExportPath, "report.md") : undefined,
    };

    const generator = new HtmlReportGenerator(
      trafficValidationResult,
      operationCoverageResult,
      0,
      options
    );
    await generator.generateHtmlReport();
  }

  private convertRuntimeException(runtimeException: RuntimeException): LiveValidationIssue {
    const ret = {
      code: runtimeException.code,
      pathsInPayload: [],
      severity: 1,
      message: runtimeException.message,
      jsonPathsInPayload: [],
      schemaPath: "",
      source: {
        url: "",
        position: {
          column: 0,
          line: 0,
        },
      },
    };

    return ret as LiveValidationIssue;
  }

  private async writeCollectionToJson(
    collectionName: string,
    collection: Collection,
    runtimeEnv: VariableScope
  ) {
    const collectionPath = path.resolve(
      this.opt.outputFolder,
      `${defaultCollectionFileName(this.opt.name, this.opt.runId!)}`
    );
    const envPath = path.resolve(
      this.opt.outputFolder,
      `${defaultEnvFileName(this.opt.name, this.opt.runId!)}`
    );
    const env = runtimeEnv.toJSON();
    env.name = collectionName + ".env";
    env._postman_variable_scope = "environment";
    await this.fileLoader.writeFile(envPath, JSON.stringify(env, null, 2));
    await this.fileLoader.writeFile(collectionPath, JSON.stringify(collection.toJSON(), null, 2));

    const values: string[] = [];
    for (const [k, v] of Object.entries(runtimeEnv.syncVariablesTo())) {
      if (this.dataMasker.maybeSecretKey(k)) {
        values.push(v as string);
      }
    }
    this.dataMasker.addMaskedValues(values);

    logger.info(`generate collection successfully!`);
    logger.info(`Postman collection: ${collectionPath}`);
    logger.info(`Postman env: ${envPath}`);
  }

  private async runCollection(runOptions: NewmanRunOptions) {
    const newmanRun = async () =>
      new Promise<NewmanRunSummary>((resolve, reject) => {
        newman.run(runOptions, function (err, summary) {
          if (summary.run.failures.length > 0) {
            process.exitCode = 1;
          }
          if (err) {
            logger.error(`collection run failed. ${err}`);
            reject(err);
          } else {
            logger.info("collection run complete!");
            resolve(summary);
          }
        });
      });
    return await newmanRun();
  }

  private async postRun(
    scenario: Scenario,
    reportExportPath: string,
    runtimeEnv: VariableScope,
    summary: NewmanRunSummary
  ) {
    const keys = await this.swaggerAnalyzer.getAllSecretKey();
    const values: string[] = [];
    for (const [k, v] of Object.entries(runtimeEnv.syncVariablesTo())) {
      if (this.dataMasker.maybeSecretKey(k)) {
        values.push(v as string);
      }
    }
    this.dataMasker.addMaskedValues(values);
    this.dataMasker.addMaskedKeys(keys);

    // add mask environment secret value
    for (const item of summary.environment.values.members) {
      if (this.dataMasker.maybeSecretKey(item.key)) {
        this.dataMasker.addMaskedValues([item.value]);
      }
    }

    const newmanReport = parseNewmanSummary(summary as any);

    const newmanReportValidatorOption: NewmanReportValidatorOption = {
      apiScenarioFilePath: scenario._scenarioDef._filePath,
      swaggerFilePaths: scenario._scenarioDef._swaggerFilePaths,
      reportOutputFilePath: defaultQualityReportFilePath(reportExportPath),
      checkUnderFileRoot: false,
      eraseXmsExamples: false,
      eraseDescription: false,
      markdown: this.opt.markdown,
      junit: this.opt.junit,
      html: this.opt.html,
      armEndpoint: this.opt.env.armEndpoint,
      runId: this.opt.runId,
      skipValidation: this.opt.skipValidation,
      generateExample: this.opt.generateExample,
      savePayload: this.opt.savePayload,
      verbose: this.opt.verbose,
    };

    const reportValidator = inversifyGetInstance(
      NewmanReportValidator,
      newmanReportValidatorOption
    );

    await reportValidator.initialize(scenario);

    await reportValidator.generateReport(newmanReport);

    if (this.opt.skipCleanUp) {
      logger.warn(
        `Notice:the resource group '${runtimeEnv.get("resourceGroupName")}' was not cleaned up.`
      );
    }
  }
}
