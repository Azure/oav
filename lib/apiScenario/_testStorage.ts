import { inversifyGetInstance } from "../inversifyUtils";
import "reflect-metadata";
import { PostmanCollectionGenerator } from "./postmanCollectionGenerator";

// oav generate-postmanCollection testScenario --env
const main = async () => {
  const env = {
    subscriptionId: "db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    location: "eastasia",
    scope: "subscriptions/db5eb68e-73e2-4fa8-b18a-46cd1be4cce5",
    resourceId: "mysourceid",
    tenantId: "<your tenantId>",
    client_id: "<your client_id>",
    client_secret: "<your client secret>",
    resourceGroupName: "ruowan-test",
    deploymentName: "createVMSS",
  };

  const postmanCollectionGenerator = inversifyGetInstance(PostmanCollectionGenerator, {
    name: "storage",
    fileRoot: "/home/ruowan/work/azure-rest-api-specs/specification/storage/resource-manager",
    swaggerFilePaths: [
      "Microsoft.Storage/stable/2019-06-01/storage.json",
      "Microsoft.Storage/stable/2019-06-01/blob.json",
    ],
    testDef: "Microsoft.Storage/stable/2019-06-01/test-scenairos/createStorageAccount.yaml",
    env: env,
    outputFolder: "./fixtures",
    generateCollection: true,
    runCollection: false,
  });

  try {
    await postmanCollectionGenerator.GenerateCollection();
  } catch (e) {
    console.log(e.message, e.stack);
  }
};

console.time("TestLoad");
console.log("Start");

main().finally(() => {
  console.timeEnd("TestLoad");
});
