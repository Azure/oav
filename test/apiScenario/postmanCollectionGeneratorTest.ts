import { PostmanCollectionGenerator } from "../../lib/apiScenario/postmanCollectionGenerator";
import { inversifyGetInstance } from "../../lib/inversifyUtils";
import { usePsudorandom } from "../../lib/util/utils";

usePsudorandom.flag = true;

describe("postmanCollectionGenerator", () => {
  it("should generate PostmanCollection - storageQuickStart", async () => {
    const generator = inversifyGetInstance(PostmanCollectionGenerator, {
      name: "storageQuickStart",
      scenarioDef: "Microsoft.Storage/stable/2021-09-01/scenarios/storageQuickStart.yaml",
      fileRoot: "test/apiScenario/fixtures/specification/storage/resource-manager/",
      checkUnderFileRoot: false,
      generateCollection: true,
      runCollection: false,
      env: {
        subscriptionId: "00000000-0000-0000-0000-000000000000",
        location: "westus",
      },
      swaggerFilePaths: ["Microsoft.Storage/stable/2021-09-01/storage.json"],
      outputFolder: "generated",
      runId: "jestRunId",
    });
    const collection = await generator.GenerateCollection();
    expect(collection).toMatchSnapshot();
  });
});
