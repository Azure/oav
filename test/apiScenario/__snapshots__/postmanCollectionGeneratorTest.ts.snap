// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postmanCollectionGenerator should generate PostmanCollection - storageQuickStart 1`] = `
Array [
  Object {
    "auth": Object {
      "bearer": Array [
        Object {
          "key": "token",
          "type": "string",
          "value": "{{bearer_token}}",
        },
      ],
      "type": "bearer",
    },
    "event": Array [
      Object {
        "id": "1",
        "listen": "prerequest",
        "script": Object {
          "exec": Array [
            "if (pm.variables.get(\\"enable_auth\\") !== \\"true\\") {",
            "    console.log(\\"Auth disabled\\");",
            "    return;",
            "}",
            "let vars = ['client_id', 'client_secret', 'tenantId', 'subscriptionId'];",
            "vars.forEach(function (item, index, array) {",
            "    pm.expect(pm.variables.get(item), item + \\" variable not set\\").to.not.be.undefined;",
            "    pm.expect(pm.variables.get(item), item + \\" variable not set\\").to.not.be.empty; ",
            "});",
            "",
            "if (!pm.collectionVariables.get(\\"bearer_token\\") || Date.now() > new Date(pm.collectionVariables.get(\\"bearer_token_expires_on\\") * 1000)) {",
            "    pm.sendRequest({",
            "        url: 'https://login.microsoftonline.com/' + pm.variables.get(\\"tenantId\\") + '/oauth2/token',",
            "        method: 'POST',",
            "        header: 'Content-Type: application/x-www-form-urlencoded',",
            "        body: {",
            "            mode: 'urlencoded',",
            "            urlencoded: [",
            "                { key: \\"grant_type\\", value: \\"client_credentials\\", disabled: false },",
            "                { key: \\"client_id\\", value: pm.variables.get(\\"client_id\\"), disabled: false },",
            "                { key: \\"client_secret\\", value: pm.variables.get(\\"client_secret\\"), disabled: false },",
            "                { key: \\"resource\\", value: \\"https://management.azure.com\\", disabled: false }",
            "            ]",
            "        }",
            "    }, function (err, res) {",
            "        if (err) {",
            "            console.log(err);",
            "        } else {",
            "            let resJson = res.json();",
            "            pm.collectionVariables.set(\\"bearer_token_expires_on\\", resJson.expires_on);",
            "            pm.collectionVariables.set(\\"bearer_token\\", resJson.access_token);",
            "        }",
            "    });",
            "}",
          ],
          "id": "2",
          "type": "text/javascript",
        },
      },
    ],
    "info": Object {
      "_postman_id": "jestRunId",
      "description": Object {
        "content": "{\\"apiScenarioFilePath\\":\\"Microsoft.Storage/stable/2021-09-01/scenarios/storageQuickStart.yaml\\",\\"apiScenarioName\\":\\"storageQuickStart_0\\",\\"swaggerFilePaths\\":[\\"Microsoft.Storage/stable/2021-09-01/storage.json\\"]}",
        "type": "text/plain",
      },
      "name": "storageQuickStart",
      "schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
      "version": undefined,
    },
    "item": Array [
      Object {
        "description": "{\\"type\\":\\"prepare\\"}",
        "event": Array [
          Object {
            "id": "5",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "pm.test(\\"response status code assertion.\\", function() {pm.response.to.be.success;});",
              ],
              "id": "6",
              "type": "text/javascript",
            },
          },
        ],
        "id": "4",
        "name": "createResourceGroup",
        "request": Object {
          "body": Object {
            "mode": "raw",
            "raw": "{\\"location\\":\\"{{location}}\\"}",
          },
          "header": Array [
            Object {
              "key": "Content-Type",
              "value": "application/json",
            },
          ],
          "method": "PUT",
          "url": Object {
            "host": Array [
              "management",
              "azure",
              "com",
            ],
            "path": Array [
              "subscriptions",
              "{{subscriptionId}}",
              "resourcegroups",
              "{{resourceGroupName}}",
            ],
            "protocol": "https",
            "query": Array [
              Object {
                "key": "api-version",
                "value": "2020-06-01",
              },
            ],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"simple\\",\\"operationId\\":\\"StorageAccounts_CheckNameAvailability\\",\\"itemName\\":\\"checkName\\",\\"step\\":\\"checkName\\"}",
        "event": Array [
          Object {
            "id": "8",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "pm.test(\\"response status code assertion.\\", function() {pm.response.to.be.success;});",
              ],
              "id": "9",
              "type": "text/javascript",
            },
          },
        ],
        "id": "7",
        "name": "checkName",
        "request": Object {
          "body": Object {
            "mode": "raw",
            "raw": "{
  \\"name\\": \\"{{accountName}}\\",
  \\"type\\": \\"Microsoft.Storage/storageAccounts\\"
}",
          },
          "header": Array [
            Object {
              "key": "Content-Type",
              "value": "application/json",
            },
          ],
          "method": "POST",
          "name": "checkName",
          "url": Object {
            "host": Array [
              "https://management",
              "azure",
              "com",
            ],
            "path": Array [
              "subscriptions",
              ":subscriptionId",
              "providers",
              "Microsoft.Storage",
              "checkNameAvailability",
            ],
            "query": Array [
              Object {
                "key": "api-version",
                "value": "2021-09-01",
              },
            ],
            "variable": Array [
              Object {
                "key": "subscriptionId",
                "type": "any",
                "value": "{{subscriptionId}}",
              },
            ],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"LRO\\",\\"poller_item_name\\":\\"createStorageAccount_poller\\",\\"operationId\\":\\"StorageAccounts_Create\\",\\"itemName\\":\\"createStorageAccount\\",\\"step\\":\\"createStorageAccount\\"}",
        "event": Array [
          Object {
            "id": "11",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "pm.test(\\"response status code assertion.\\", function() {pm.response.to.be.success;});",
              ],
              "id": "12",
              "type": "text/javascript",
            },
          },
          Object {
            "id": "13",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "",
                "        const pollingUrl = pm.response.headers.get('Location') || pm.response.headers.get('Azure-AsyncOperation');",
                "        if (pollingUrl) {",
                "          pm.collectionVariables.set(\\"x_polling_url\\", pollingUrl);",
                "      }",
                "        ",
              ],
              "id": "14",
              "type": "text/javascript",
            },
          },
        ],
        "id": "10",
        "name": "createStorageAccount",
        "request": Object {
          "body": Object {
            "mode": "raw",
            "raw": "{
  \\"sku\\": {
    \\"name\\": \\"Standard_LRS\\"
  },
  \\"kind\\": \\"StorageV2\\",
  \\"location\\": \\"{{location}}\\",
  \\"properties\\": {
    \\"encryption\\": {
      \\"services\\": {
        \\"blob\\": {}
      },
      \\"keySource\\": \\"Microsoft.Storage\\"
    }
  }
}",
          },
          "header": Array [
            Object {
              "key": "Content-Type",
              "value": "application/json",
            },
          ],
          "method": "PUT",
          "name": "createStorageAccount",
          "url": Object {
            "host": Array [
              "https://management",
              "azure",
              "com",
            ],
            "path": Array [
              "subscriptions",
              ":subscriptionId",
              "resourceGroups",
              ":resourceGroupName",
              "providers",
              "Microsoft.Storage",
              "storageAccounts",
              ":accountName",
            ],
            "query": Array [
              Object {
                "key": "api-version",
                "value": "2021-09-01",
              },
            ],
            "variable": Array [
              Object {
                "key": "resourceGroupName",
                "type": "any",
                "value": "{{resourceGroupName}}",
              },
              Object {
                "key": "accountName",
                "type": "any",
                "value": "{{accountName}}",
              },
              Object {
                "key": "subscriptionId",
                "type": "any",
                "value": "{{subscriptionId}}",
              },
            ],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"poller\\",\\"lro_item_name\\":\\"createStorageAccount\\"}",
        "event": Array [
          Object {
            "id": "19",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "",
                "      try{",
                "        if(pm.response.code === 202){",
                "          postman.setNextRequest('_createStorageAccount_poller_mock_delay')",
                "        }else if(pm.response.code === 204){",
                "          postman.setNextRequest('_createStorageAccount_generated_get')",
                "        }",
                "        else{",
                "          const terminalStatus = [\\"Succeeded\\", \\"Failed\\", \\"Canceled\\"]",
                "          if(pm.response.json().status !== undefined && terminalStatus.indexOf(pm.response.json().status) === -1){",
                "            postman.setNextRequest('_createStorageAccount_poller_mock_delay')",
                "          }else{",
                "            postman.setNextRequest('_createStorageAccount_generated_get')",
                "          }",
                "        }",
                "      }catch(err){",
                "        postman.setNextRequest('_createStorageAccount_generated_get')",
                "      }",
              ],
              "id": "35",
              "type": "text/javascript",
            },
          },
        ],
        "id": "15",
        "name": "_createStorageAccount_poller",
        "request": Object {
          "method": "GET",
          "url": Object {
            "host": Array [
              "{{x_polling_url}}",
            ],
            "query": Array [],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"mock\\",\\"lro_item_name\\":\\"createStorageAccount\\"}",
        "event": Array [
          Object {
            "id": "17",
            "listen": "prerequest",
            "script": Object {
              "exec": Array [
                "postman.setNextRequest('_createStorageAccount_poller')",
              ],
              "id": "18",
              "type": "text/javascript",
            },
          },
        ],
        "id": "16",
        "name": "_createStorageAccount_poller_mock_delay",
        "request": Object {
          "method": "GET",
          "url": Object {
            "host": Array [
              "postman-echo",
              "com",
            ],
            "path": Array [
              "delay",
              "10",
            ],
            "protocol": "https",
            "query": Array [],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"generated-get\\",\\"lro_item_name\\":\\"createStorageAccount\\",\\"step\\":\\"createStorageAccount\\"}",
        "event": Array [
          Object {
            "id": "22",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "pm.test(\\"response status code assertion.\\", function() {pm.response.to.be.success;});",
              ],
              "id": "23",
              "type": "text/javascript",
            },
          },
        ],
        "id": "21",
        "name": "_createStorageAccount_generated_get",
        "request": Object {
          "header": Array [
            Object {
              "key": "Content-Type",
              "value": "application/json",
            },
          ],
          "method": "GET",
          "url": Object {
            "host": Array [
              "management",
              "azure",
              "com",
            ],
            "path": Array [
              "subscriptions",
              "{{subscriptionId}}",
              "resourceGroups",
              "{{resourceGroupName}}",
              "providers",
              "Microsoft.Storage",
              "storageAccounts",
              "{{accountName}}",
            ],
            "protocol": "https",
            "query": Array [
              Object {
                "key": "api-version",
                "value": "2021-09-01",
              },
            ],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "event": Array [
          Object {
            "id": "25",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "pm.test(\\"response code should be 2xx\\", function() {pm.response.to.be.success;});",
              ],
              "id": "26",
              "type": "text/javascript",
            },
          },
          Object {
            "id": "27",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "",
                "        const pollingUrl = pm.response.headers.get('Location') || pm.response.headers.get('Azure-AsyncOperation');",
                "        if (pollingUrl) {",
                "          pm.collectionVariables.set(\\"x_polling_url\\", pollingUrl);",
                "      }",
                "        ",
              ],
              "id": "28",
              "type": "text/javascript",
            },
          },
        ],
        "id": "24",
        "name": "deleteResourceGroup",
        "request": Object {
          "header": Array [
            Object {
              "key": "Content-Type",
              "value": "application/json",
            },
          ],
          "method": "DELETE",
          "url": Object {
            "host": Array [
              "management",
              "azure",
              "com",
            ],
            "path": Array [
              "subscriptions",
              "{{subscriptionId}}",
              "resourcegroups",
              "{{resourceGroupName}}",
            ],
            "protocol": "https",
            "query": Array [
              Object {
                "key": "api-version",
                "value": "2020-06-01",
              },
            ],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"poller\\",\\"lro_item_name\\":\\"deleteResourceGroup\\"}",
        "event": Array [
          Object {
            "id": "33",
            "listen": "test",
            "script": Object {
              "exec": Array [
                "",
                "      try{",
                "        if(pm.response.code === 202){",
                "          postman.setNextRequest('_deleteResourceGroup_poller_mock_delay')",
                "        }else if(pm.response.code === 204){",
                "          postman.setNextRequest(null)",
                "        }",
                "        else{",
                "          const terminalStatus = [\\"Succeeded\\", \\"Failed\\", \\"Canceled\\"]",
                "          if(pm.response.json().status !== undefined && terminalStatus.indexOf(pm.response.json().status) === -1){",
                "            postman.setNextRequest('_deleteResourceGroup_poller_mock_delay')",
                "          }else{",
                "            postman.setNextRequest(null)",
                "          }",
                "        }",
                "      }catch(err){",
                "        postman.setNextRequest(null)",
                "      }",
              ],
              "id": "36",
              "type": "text/javascript",
            },
          },
        ],
        "id": "29",
        "name": "_deleteResourceGroup_poller",
        "request": Object {
          "method": "GET",
          "url": Object {
            "host": Array [
              "{{x_polling_url}}",
            ],
            "query": Array [],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
      Object {
        "description": "{\\"type\\":\\"mock\\",\\"lro_item_name\\":\\"deleteResourceGroup\\"}",
        "event": Array [
          Object {
            "id": "31",
            "listen": "prerequest",
            "script": Object {
              "exec": Array [
                "postman.setNextRequest('_deleteResourceGroup_poller')",
              ],
              "id": "32",
              "type": "text/javascript",
            },
          },
        ],
        "id": "30",
        "name": "_deleteResourceGroup_poller_mock_delay",
        "request": Object {
          "method": "GET",
          "url": Object {
            "host": Array [
              "postman-echo",
              "com",
            ],
            "path": Array [
              "delay",
              "10",
            ],
            "protocol": "https",
            "query": Array [],
            "variable": Array [],
          },
        },
        "response": Array [],
      },
    ],
    "variable": Array [
      Object {
        "key": "subscriptionId",
        "type": "any",
      },
      Object {
        "key": "resourceGroupName",
        "type": "any",
      },
      Object {
        "key": "location",
        "type": "any",
      },
      Object {
        "key": "enable_auth",
        "type": "any",
        "value": "true",
      },
      Object {
        "key": "client_id",
        "type": "any",
      },
      Object {
        "key": "client_secret",
        "type": "any",
      },
      Object {
        "key": "tenantId",
        "type": "any",
      },
      Object {
        "key": "bearer_token",
        "type": "any",
      },
      Object {
        "key": "bearer_token_expires_on",
        "type": "any",
      },
    ],
  },
]
`;
